pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -255,97 +179,93 @@ func (s *eventStream) Close(ctx context.Context) {
 		s.rf.Close()
 	}
 	s.acc.Close(ctx)
-	if s.doneChan != nil {
-		close(s.doneChan)
-	}
-	if err := s.streamGroup.Wait(); err != nil {
-		// Note: error in close is normal; we expect to be terminated with context canceled.
-		log.Errorf(ctx, "partition stream %d terminated with error %v", s.streamID, err)
-	}
-	if s.sp != nil {
-		s.sp.Finish()
-	}
-}
-
-func (s *eventStream) onValue(ctx context.Context, value *kvpb.RangeFeedValue) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{Val: value}}:
-		log.VInfof(ctx, 1, "onValue: %s@%s", value.Key, value.Value.Timestamp)
-	}
 }
 
-func (s *eventStream) onCheckpoint(ctx context.Context, checkpoint *kvpb.RangeFeedCheckpoint) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{Checkpoint: checkpoint}}:
-		log.VInfof(ctx, 1, "onCheckpoint: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
-	}
+func (s *eventStream) onInitialScanDone(ctx context.Context) {
+	// We no longer expect concurrent onValue calls so we can remove the mu.
+	s.addMu = nil
 }
 
-func (s *eventStream) onInitialScanSpanCompleted(ctx context.Context, sp roachpb.Span) error {
-	checkpoint := kvpb.RangeFeedCheckpoint{
-		Span:       sp,
-		ResolvedTS: s.spec.InitialScanTimestamp,
-	}
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{Checkpoint: &checkpoint},
-	}:
-		log.VInfof(ctx, 1, "onSpanCompleted: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
-		return nil
-	}
+func (s *eventStream) onValue(ctx context.Context, value *kvpb.RangeFeedValue) {
+	// During initial-scan we expect concurrent onValue calls from the parallel
+	// scan workers, but once the initial scan ends the mu will be nilled out and
+	// we can avoid the locking overhead here.
+	if s.addMu != nil {
+		s.addMu.Lock()
+		defer s.addMu.Unlock()
+	}
+	s.seb.addKV(roachpb.KeyValue{Key: value.Key, Value: value.Value})
+	s.setErr(s.maybeFlushBatch(ctx))
 }
 
 func (s *eventStream) onSSTable(
 	ctx context.Context, sst *kvpb.RangeFeedSSTable, registeredSpan roachpb.Span,
 ) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{SST: sst},
-		RegisteredSpan: registeredSpan,
-	}:
-		log.VInfof(ctx, 1, "onSSTable: %s@%s with registered span %s",
-			sst.Span, sst.WriteTS, registeredSpan)
+	if s.setErr(s.addSST(sst, registeredSpan)) {
+		return
 	}
+	s.setErr(s.maybeFlushBatch(ctx))
```

Comment:
so, we don't grab the lock for onSSTable and onDeleteRange because they're never called during initial scan? And you were only able to figure this out via trial and error with CI? Insider trading is sad.

Reply:
If onSSTable isn't used during initial scan, I wonder if we could speed the init scan up by pushing addSStable requests...

Reply:
> And you were only able to figure this out via trial and error with CI?
no, not via trial and error: the initial scan path is very tractable and is only passed `rowsFn`

Reply:
FWIW, there is a comment in rangefeed.go that says _all_ the callbacks are called in serial, so I think it is ok to treat that as the default; the fact that onValue and checkpoint get called is special, and why they get the mu